site/js/util/modal.js

```diff
@@ -0,0 +1,131 @@
+// pretty Modal class. replaces window.prompt and friends
+
+const Nanobus = require('nanobus')
+const html = require('choo/html')
+
+const constructStyledInput = (name, i, j) => {
+  const id = `modal-input-${name}`
+
+  return html`<div class='styledInput'>
+    <label for=${id}>${i.label}</label>
+    <input id=${id} type=${i.type || 'text'} placeholder=${i.placeholder || ''} tabindex=${j}/>
+  </div>`
+}
+
+class Modal extends Nanobus {
+  constructor(opts) {
+    super('modal')
+
+    const content = html`<div class='modal-content'></div>`
+
+    // add inputs to content
+    let j = 1
+    this.styledInputs = []
+    for (const [ name, i ] of Object.entries(opts.inputs)) {
+      const inputEl = constructStyledInput(name, i, j++)
+
+      this.styledInputs.push({ name, el: inputEl })
+      content.appendChild(inputEl)
+    }
+
+    // hitting return on an input element should focus
+    // the next, or submit the modal if it is last
+    for (let i = 0; i < this.styledInputs.length; i++) {
+      const { name, el } = this.styledInputs[i]
+      const next = this.styledInputs[i + 1]
+      const input = el.querySelector('input')
+
+      input.addEventListener('keypress', evt => {
+        // listen for enter/return keypress
+        if (evt.which === 13) {
+          evt.preventDefault()
+
+          // if there's a next input, focus it
+          // otherwise, submit this modal
+          if (next) {
+            next.el.querySelector('input').focus()
+          } else {
+            this.submit()
+          }
+        }
+      })
+    }
+
+    // add submit button to content
+    const btn = html`<input type='submit' class='styledButton' value=${opts.button || 'Submit'} onclick=${this.submit.bind(this)} tabindex=${j}>`
+    content.appendChild(btn)
+
+    // add error element
+    // TODO style this better! this is a feature PJ's codepen didn't have
+    this.errorEl = html`<div class='modal-error'></div>`
+    content.prepend(this.errorEl)
+
+    // construct #modal element
+    this.el = html`<div class='modal'>
+      <div class='modal-close-button' onclick=${this.close.bind(this)}></div>
+      <div class='modal-header'>${opts.title} <span class="modal-header-subtitle">${opts.subtitle || ''}</span></div>
+      ${content}
+    </div>`
+
+    // darken page cover
+    const pageCover = document.querySelector('.modal-page-cover')
+    pageCover.onclick = this.close.bind(this) // close on click
+    pageCover.classList.add('visible')
+
+    // show the modal
+    this.visible = true
+    document.body.insertBefore(this.el, pageCover)
```

Comment:
The modal element goes *before* the page cover?

Reply:
Yep.

Reply:
Why?

Reply:
For some reason PullJosh's CSS only works when it's inserted before the page cover. It doesn't really matter, though.